tigris/model.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2022 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tigris
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/google/uuid"
+	api "github.com/tigrisdata/tigris-client-go/api/server/v1"
+)
+
+// User document model can have its own implementation
+// of this interface to customize metadata handling
+type model interface {
+	SetCreatedAt(t time.Time)
+	SetUpdatedAt(t time.Time)
+	SetDeletedAt(t time.Time)
+}
+
+// Model contains document metadata
+// Metadata is populated automatically if embedded into user document
+//
+// Example:
+//    type User {
+//       tigris.Model
+//       // User fields
+//       Name string
+//       ...
+//    }
+type Model struct {
+	// These fields are automatically populated from
+	// request metadata
+	CreatedAt time.Time `json:"-"`
+	UpdatedAt time.Time `json:"-"`
+	DeletedAt time.Time `json:"-"`
+
+	// This field is unused if user defines its own primary key
+	// If no primary key is defined in the user model
+	// this field is automatically populated with
+	// autogenerated unique UUID
+	ID uuid.UUID `json:"id" tigris:"primary_key,generate"`
```

Comment:
ah, interesting. Let's coordinate with Jigar as well then this is a cool extension. Also, let's document this.

Reply:
`autogenerate `

Reply:
That's not used at the moment, schema support will be added in the next diff.

Reply:
that's fine, we can still use the same key name.